### PR TITLE
[SeparateConstOffsetFromGEP] Reorder trivial GEP chains to separate constants

### PR DESCRIPTION
**Pull request by jrbyrnes**
In this case, a trivial GEP chain has the form:

```
%ptr = getelementptr sameType, %base, constant
%val = getelementptr sameType, %ptr, %variable
```

That is, a one-index GEP consumes another (of the same basis and result type) one-index GEP, where the inner GEP uses a constant index and the outer GEP uses a variable index. For chains of this type, it is trivial to reorder them (by simply swapping the indexes). The result of doing so is better AddrMode matching for users of the ultimate ptr produced by GEP chain.

Future patches can extend this to support non-trivial GEP chains (e.g. those with different basis types and/or multiple indices).

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp [hunk 1]
```diff
@@ -987,11 +987,50 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
...
+                                     /*BaseGV=*/nullptr, NestedByteOffset,
+                                     /*HasBaseReg=*/true, /*Scale=*/0,
+                                     AddrSpace)) {
+        return Changed;
+      }
+
+      IRBuilder<> Builder(GEP);
+      Builder.SetCurrentDebugLocation(GEP->getDebugLoc());
+      // For trivial GEP chains, we can swap the indicies.
+      auto NewSrc = Builder.CreateGEP(PtrGEPType, PtrGEP->getPointerOperand(),
```
Comment (arsenm) [inline llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:1023]: Can this preserve inbounds? Can you add some mixed inbounds usage tests?

File: llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp [hunk 2]
```diff
@@ -987,11 +987,50 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
...
+    // Attempt to reassociate GEPs for better AddrMode construction.
+    if (auto PtrGEP = dyn_cast<GetElementPtrInst>(GEP->getPointerOperand())) {
+      // TODO: support reordering for non-trivial GEP chains
+      if (PtrGEP->getNumIndices() != 1 || GEP->getNumIndices() != 1)
+        return Changed;
+
+      Type *GEPType = GEP->getResultElementType();
+      Type *PtrGEPType = PtrGEP->getResultElementType();
+      // TODO: support reordering for non-trivial GEP chains
+      if ((PtrGEPType != GEPType) ||
```
Comment (arsenm) [inline llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:1002]: Remove parentheses around the compare

File: llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp [hunk 3]
```diff
@@ -987,11 +987,50 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
...
   int64_t AccumulativeByteOffset = accumulateByteOffset(GEP, NeedsExtraction);
 
-  if (!NeedsExtraction)
-    return Changed;
-
   TargetTransformInfo &TTI = GetTTI(*GEP->getFunction());
 
+  if (!NeedsExtraction) {
+    // Attempt to reassociate GEPs for better AddrMode construction.
+    if (auto PtrGEP = dyn_cast<GetElementPtrInst>(GEP->getPointerOperand())) {
```
Comment (arsenm) [inline llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:994]: Reduce the indentation over the GEP block? Could either separate to a separate function or use C++17 if statement initializers

**Review by arsenm — COMMENTED** [review]
I'm somewhat confused by the description, since this is what I thought the point of the pass was and it already did. What exactly is special about this particular case?

**Comment by jrbyrnes** [comment]
> What exactly is special about this particular case?

When analyzing the GEP operands for constants, this pass only supports certain Value classes: {ConstantInt, BinaryOperator, TruncInst, SExtInst, ZExtInst}. In this case, the operand is a GetElementPtrInst, so the pass does not attempt to reorganize the GEP in such a way that the constant is the outtermost index.

I think it is feasible to extend the set of supported Value classes to include GEPs (primarily since we can assume nested GEPs will already have their constants extracted), but that requires a more significant rework. This feature can bridge the gap until then.

**Review by krzysz00 — APPROVED** [review]
Looks good to me!

**Comment by arsenm** [comment]
Can you add an alive2 link that shows the inbounds preserve is OK?

**Comment by arsenm** [comment]
reverse ping

**Comment by jrbyrnes** [comment]
So it appears that the inbounds preservation is not correct.

Preserved:
https://alive2.llvm.org/ce/z/fdH9f2

Stripped:
https://alive2.llvm.org/ce/z/dJwv-W

Instead of stripping the inbounds, I've disabled the optimization for inbounds GEPs

**Comment by krzysz00** [comment]
... hold on, from looking at that Alive2, I think you can be a bit more aggressive than just not applying this swap to inbounds GEPs (or stripping the inbounds - do we know which of these is more likely to lead to good codegen?).

Specifically, I claim that, if you can prove that the constant and variable indices are non-negative, you should be able to reassociate the inbounds GEPs soundly. 

See the somewhat ham-fisted https://alive2.llvm.org/ce/z/CRkgEN

**Comment by krzysz00** [comment]
Do you need someone to land this or?

**Comment by jrbyrnes** [comment]
> Do you need someone to land this or?

I was just waiting to see if there were any objections with the newest version of the changes -- based on this, I'll assume not. Thanks for the review!

**Comment by jrbyrnes** [comment]
I guess some unaccounted for change had come along to effect the behavior of NVPTX/lower-gep-reorder.ll

I've resolved the check-llvm fails via https://github.com/llvm/llvm-project/commit/ec0aa1646e9953d1a8d0d15dc381d3250c854572

**Comment by nico** [comment]
Looks like this still breaks tests: http://45.33.8.238/linux/130717/step_12.txt

This is after ec0aa1646e9953d1a8d0d15dc381d3250c854572.

Should we revert for now?

**Comment by krzysz00** [comment]
Might be good to revert and make sure there's a local test run with all targets enabled at least?

**Comment by AaronHLiu** [comment]
Looks like there are some LIT failures need to be addressed:
```
FAIL: LLVM :: Transforms/SeparateConstOffsetFromGEP/NVPTX/lower-gep-reorder.ll (68679 of 78188)
******************** TEST 'LLVM :: Transforms/SeparateConstOffsetFromGEP/NVPTX/lower-gep-reorder.ll' FAILED ********************
Exit Code: 1
```

**Comment by AaronHLiu** [comment]
```
; CHECK-NEXT: [[TMP1:%.*]] = getelementptr i64, ptr [[TMP0]], i64 [[IN_IDX1]]
              ^
<stdin>:9:67: note: scanning from here
 %0 = getelementptr [8192 x i64], ptr %in.ptr, i64 %in.idx0, i64 0
                                                                  ^
<stdin>:9:67: note: with "TMP0" equal to "%0"
 %0 = getelementptr [8192 x i64], ptr %in.ptr, i64 %in.idx0, i64 0
                                                                  ^
<stdin>:9:67: note: with "IN_IDX1" equal to "%in\\.idx1"
 %0 = getelementptr [8192 x i64], ptr %in.ptr, i64 %in.idx0, i64 0
                                                                  ^
<stdin>:10:6: note: possible intended match here
 %const11 = getelementptr i8, ptr %0, i64 2048
     ^

Input file: <stdin>
Check file: /home/hualilia/llvm/dev/llvm-project/llvm/test/Transforms/SeparateConstOffsetFromGEP/NVPTX/lower-gep-reorder.ll

-dump-input=help explains the following input dump.

Input was:
<<<<<<
           1: ; ModuleID = '<stdin>'
           2: source_filename = "<stdin>"
           3: target datalayout = "e-i64:64-i128:128-v16:16-v32:32-n16:32:64"
           4: target triple = "nvptx64-nvidia-cuda"
           5:
           6: define protected amdgpu_kernel void @sink_addr(ptr %in.ptr, i64 %in.idx0, i64 %in.idx1) {
           7: entry:
           8:  %idx0 = getelementptr [8192 x i64], ptr %in.ptr, i64 %in.idx0, i64 %in.idx1
           9:  %0 = getelementptr [8192 x i64], ptr %in.ptr, i64 %in.idx0, i64 0
next:10'0                                                                       X error: no match found
next:10'1                                                                         with "TMP0" equal to "%0"
next:10'2                                                                         with "IN_IDX1" equal to "%in\\.idx1"
          10:  %const11 = getelementptr i8, ptr %0, i64 2048
next:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:10'1                                                                         with "TMP0" equal to "%0"
next:10'2                                                                         with "IN_IDX1" equal to "%in\\.idx1"
          10:  %const11 = getelementptr i8, ptr %0, i64 2048
next:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:10'3          ?                                          possible intended match
          11:  %idx1 = getelementptr i64, ptr %const11, i64 %in.idx1
next:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          12:  %1 = getelementptr [8192 x i64], ptr %in.ptr, i64 %in.idx0, i64 0
next:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          13:  %const22 = getelementptr i8, ptr %1, i64 4096
next:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          14:  %idx2 = getelementptr i64, ptr %const22, i64 %in.idx1
next:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          15:  %2 = getelementptr [8192 x i64], ptr %in.ptr, i64 %in.idx0, i64 0
next:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           .
```

**Comment by krzysz00** [comment]
... that got handled in the re-land?

**Comment by jrbyrnes** [comment]
@AaronHLiu this PR was known to cause that exact LIT failure -- this was why it was reverted upstream -- does your build have the revert?

The reland should be okay; in you deleted comment there was a CHECK line flagged that exist in that bad version of the LIT test and not the reland version.

**Comment by AaronHLiu** [comment]
> @AaronHLiu this PR was known to cause that exact LIT failure -- this was why it was reverted upstream -- does your build have the revert?
> 
> The reland should be okay; in you deleted comment there was a CHECK line flagged that existed in that bad version of the LIT test and not the reland version (reland: #81671) .

The build did not contain the new version yet. Will check. Thanks!

**Pull request merged**